provider/aws.go

```diff
@@ -407,60 +407,74 @@ func (p *AWSProvider) newChanges(action string, endpoints []*endpoint.Endpoint,
 	changes := make([]*route53.Change, 0, len(endpoints))
 
 	for _, endpoint := range endpoints {
-		changes = append(changes, p.newChange(action, endpoint, recordsCache, zones))
+		change, dualstack := p.newChange(action, endpoint, recordsCache, zones)
+		changes = append(changes, change)
+		if dualstack {
+			// make a copy of change, modify RRS type to AAAA, then add new change
+			rrs := *change.ResourceRecordSet
+			change2 := &route53.Change{Action: change.Action, ResourceRecordSet: &rrs}
+			change2.ResourceRecordSet.Type = aws.String(route53.RRTypeAaaa)
+			changes = append(changes, change2)
+		}
 	}
 
 	return changes
 }
 
-// newChange returns a Change of the given record by the given action, e.g.
+// newChange returns a route53 Change and a boolean indicating if there should also be a change to a AAAA record
+// returned Change is based on the given record by the given action, e.g.
 // action=ChangeActionCreate returns a change for creation of the record and
 // action=ChangeActionDelete returns a change for deletion of the record.
-func (p *AWSProvider) newChange(action string, endpoint *endpoint.Endpoint, recordsCache []*endpoint.Endpoint, zones map[string]*route53.HostedZone) *route53.Change {
+func (p *AWSProvider) newChange(action string, ep *endpoint.Endpoint, recordsCache []*endpoint.Endpoint, zones map[string]*route53.HostedZone) (*route53.Change, bool) {
 	change := &route53.Change{
 		Action: aws.String(action),
 		ResourceRecordSet: &route53.ResourceRecordSet{
-			Name: aws.String(endpoint.DNSName),
+			Name: aws.String(ep.DNSName),
 		},
 	}
+	dualstack := false
 
-	if isAWSLoadBalancer(endpoint) {
+	if isAWSLoadBalancer(ep) {
 		evalTargetHealth := p.evaluateTargetHealth
-		if prop, ok := endpoint.GetProviderSpecificProperty(providerSpecificEvaluateTargetHealth); ok {
+		if prop, ok := ep.GetProviderSpecificProperty(providerSpecificEvaluateTargetHealth); ok {
 			evalTargetHealth = prop.Value == "true"
 		}
+		// If the endpoint has a Dualstack label, append a change for AAAA record as well.
+		if val, ok := ep.Labels[endpoint.DualstackLabelKey]; ok {
```

Comment:
Can you check this PR https://github.com/kubernetes-incubator/external-dns/pull/576
there was already some discussions about naming labels

Reply:
will do

Reply:
It looks to me like #576 seeks to support AAAA records in the general case, but not the concept of a "dualstack" target. I completely agree that DNS records are not dualstack (although CNAMEs do return either AAAA or A depending on the query.) But, devices can be dualstack, and the point of the ingress source is to create the appropriate DNS records that correspond to hostname(s) associated with a given ingress. So, if the ingress corresponds to a dualstack device (an ALB in this case) then I think the registered target for each hostname should be flagged as a "dualstack" target. (This is why I added the concept of a dualstack label -- to ensure that the registry captured this property of the target.)

The AWS provider replaces a CNAME target with an AWS alias record (in the case where the target represents an AWS load balancer,) I believe that the resulting (DNS query) behavior of the created alias(es) should be similar to the CNAME that these replace -- i.e. that it should create two aliases if the target is a load balancer that is "dualstack" as AWS describes it. If a query for either an A or AAAA record for a CNAME to a given target will result in either an A or AAAA answer, then the same 2 queries to alias records should give back the same answers (without the extra CNAME step.)

#576 doesn't address the specific case of a dualstack target, but rather expects the owner of the ingress resource to think about the fact that both an A and AAAA record need to be created and to add an appropriate annotation. I think this is counter to the way external-dns works in other cases. e.g. external-dns substitutes an alias target when it seems appropriate to do that instead of a CNAME for a given provider, rather than making the resource owner specify that behavior with a specific annotation. I like my approach of recognizing that the target should be treated as a dualstack target if the underlying load balancer resource is dualstack.